server/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java

```diff
@@ -274,24 +276,48 @@ public ScheduledCancellable schedule(Runnable command, TimeValue delay, Executor
     }
 
     public void testRunnableDoesNotRunAfterCancellation() throws Exception {
-        final int iterations = scaledRandomIntBetween(2, 12);
-        final AtomicInteger counter = new AtomicInteger();
-        final CountDownLatch doneLatch = new CountDownLatch(iterations);
-        final Runnable countingRunnable = () -> {
+        int iterations = scaledRandomIntBetween(2, 12);
+
+        // we don't have the cancellable until we schedule the task, which needs the barrier object to reference in the closure
+        // so break the circular dependency here
+        AtomicReference<Runnable> checkCancel = new AtomicReference<>();
+
+        AtomicInteger counter = new AtomicInteger();
+        CyclicBarrier barrier = new CyclicBarrier(2, () -> checkCancel.get().run());
```

Comment:
This is clever, but it required me a bit too much to understand why 2 and how it interacts with the `for(i.. iteration) barrier.await()`. 
I see the point of using basic primitives in this kind of tests, to be as much as possible in control and depend on as little as possible, but I wonder if using `CompletableFuture` here could bring an advantage in readability.

Reply:
I'm not sure how CompetableFuture could be used here_

Reply:
You are right, I was thinking about having `countingRunnable` complete futures from an array and then `CompletableFuture.allOf(c).whenComplete((unused, throwable) -> cancellable.get().cancel());`, but reading the docs there is no guarantee that the continuation will run on the completing thread (it might, but it's not certain), opening the door for "another round" of counting to go. While the CyclicBarrier action is " performed by the last thread entering the barrier."